### DISCUSSION TITLE
Announcement: Plugins are coming to Biome 2.0

### DISCUSSION THREAD

**Discussion by arendjr** [biomejs/biome#4428] [Announcements]
As the team is preparing for the Biome 2.0 release (no ETA yet, sorry!), I am happy to reveal one of the features that will be a part of that release...

**Initial plugin support has been merged:** https://github.com/biomejs/biome/pull/4396

This builds on top of the GritQL integration that we [shipped as part of the 1.9 release](https://biomejs.dev/blog/biome-v1-9/#search-command), and we've already merged various other performance, compatibility and syntax improvements since then.

At this point, it is important to stress that the exact feature set for plugins in the 2.0 release is not yet fixed. At the very least, users should be able to **create GritQL plugins to write custom lint rules for reporting diagnostics in JS/TS files**. 

We have quite some TODOs still on our wishlist:

* Adding support for (auto-)fixers in plugins
* https://github.com/biomejs/biome/issues/4427
* Improvements to our https://github.com/biomejs/biome/issues/2582. Specifically:
  * Support for querying comments
  * Handling suppressions
  * Better support for [querying AST nodes](https://docs.grit.io/language/patterns#syntax-tree-ast-nodes)
* Shout-out to @abidjappie and @branberry who have been helping us towards making the https://github.com/biomejs/biome/issues/2476 a reality!

And post-2.0, we'll want to continue looking into [plugins that can be written in JS/TS](https://github.com/biomejs/biome/discussions/2467) as well.

If you think you can help us out with any of these tasks, please let us know! If not, just be patient and know that good stuff is on the way :) Cheers!

**Comment by vorant94** [comment]
first on all great work you are doing! I really like biome and it improved my dev experience by a lot compared to its alternatives.

I'm curious about one thing regarding biome plugin system: how is it different to where eslint now stands?

like one of my personal pain points with eslint was that for every small thing i need some 3-rd party plugin with at best 20k npm downloads a month while biome has EVERYTHING out of the box. does biome plugin system means that in a couple of years the same will be true for biome as well?

**Reply by arendjr** [reply]
> in a couple of years the same will be true for biome as well?

We surely hope not!

We very much intent to stick to our batteries-included approach with Biome. Plugins are intended either as workaround for things we haven’t (yet) implemented, or that are too project-specific for us to include in Biome.

Generally, our stance is that if a plugin becomes popular enough that people want to share it, that likely means we'll want to upstream it and create a built-in rule for it. Those are still easier to use for our users, and they're more performant as well.

**Comment by jacob-lcs** [comment]
Wondering when version 2.0 will be available? 😁

**Reply by arendjr** [reply]
We are working hard on it, but we don’t do ETAs because it’s still a volunteer effort.

**Reply by plaa** [reply]
I read through the [Road to Biome 2.0](https://github.com/biomejs/biome/issues/3727).  May I ask why is there such a long list of features desired for the 2.0 release, instead of doing smaller releases (even with breaking changes) more often?  I understand the desire of bundling breaking changes together, but having a high bar on major releases easily leads to stagnation in open source projects.  There are a lot of cool features already implemented, is there something broken that would prevent releasing them for general use?

**Reply by ematipico** [reply]
More it's covered here: https://github.com/biomejs/biome/issues/5280#issuecomment-2703429421

We are still a relatively young team working on a relatively young project. v1 was essentially a quick fork from the old project, where we rushed in, hoping to keep the toolchain's initial mission alive.

v2, for us, is also a way to stop for a bit, take a deep breath, and reevaluate things so we don't burn out and we can reach the maturity of those projects that can ship fast, do smaller releases and break often.

**Reply by plaa** [reply]
Thank you @ematipico for the explanation. I fully appreciate taking the time to fix things and especially to take care of yourselves. I just hope you're not feeling too much pressure or setting too high a bar either quality or feature-wise when making releases. Thank you for your hard work.  ❤️
